daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Anf.scala

```diff
@@ -0,0 +1,426 @@
+        Bounce(() => txK(AExpr(sexpr)))
+    })
+  }
+
+  private[this] def transformLet1[A](
+      depth: DepthA,
+      env: Env,
+      rhs: SExpr,
+      body: SExpr,
+      k: K[AExpr, A],
+      transform: Tx[SExpr, A]): Trampoline[A] = {
+    Bounce(() =>
+      transformExp(depth, env, rhs, k) { (depth, rhs, txK) =>
+        val depth1 = depth.incr(1)
+        val env1 = trackBindings(depth, env, 1)
+        Bounce(() =>
+          transformExp(depth1, env1, body, { body1 =>
+            Bounce(() => txK(AExpr(SELet1(rhs, body1.wrapped))))
+          })(transform))
+    })
+  }
+
+  private[this] def flattenAlts[A](depth: DepthA, env: Env, alts: Array[SCaseAlt])(
+      k: K[Array[SCaseAlt], A]): Trampoline[A] = {
+    // Note: this could be made properly CPS and thus constant stack through
+    // trampoline by implementing a CPS version of map. However, map on an
+    // array is implemented as a loop so this should be fine.
+    Bounce(() =>
+      k(alts.map {
+        case SCaseAlt(pat, body0) =>
+          val n = patternNArgs(pat)
+          val env1 = trackBindings(depth, env, n)
+          flattenExp(depth.incr(n), env1, body0)(body => {
+            Land(SCaseAlt(pat, body.wrapped))
+          }).bounce
+      }))
+  }
+
+  private[this] def patternNArgs(pat: SCasePat): Int = pat match {
+    case _: SCPEnum | _: SCPPrimCon | SCPNil | SCPDefault | SCPNone => 0
+    case _: SCPVariant | SCPSome => 1
+    case SCPCons => 2
+  }
+
+  /** `transformExp` is the function at the heart of the ANF transformation.
+    *  You can read its type as saying: "Caller, give me a general expression
+    *  `exp`, (& depth/env info), and a transformation function `transform`
+    *  which says what you want to do with the transformed expression. Then I
+    *  will do the transform, and call `transform` with it. I reserve the right
+    *  to wrap further expression-AST around the expression returned by
+    *  `transform`.
+    *
+    *  See: `atomizeExp` for an instance where this wrapping occurs.
+    *
+    *  Note: this wrapping is the reason why we need a "second" CPS transform to
+    *  achieve constant stack through trampoline.
+    */
+  private[this] def transformExp[A](depth: DepthA, env: Env, exp: SExpr, k: K[AExpr, A])(
+      transform: Tx[SExpr, A]): Trampoline[A] =
+    exp match {
+      case atom0: SExprAtomic =>
+        val atom = makeRelativeA(depth)(makeAbsoluteA(env, atom0))
+        Bounce(() => transform(depth, atom, k))
+
+      case x: SEVal => Bounce(() => transform(depth, x, k))
+      case x: SEImportValue => Bounce(() => transform(depth, x, k))
+
+      case SEAppGeneral(func, args) =>
+        // It's safe to perform ANF if the func-expression has no effects when evaluated.
+        val safeFunc =
+          func match {
+            // we know that trivially in these two cases
+            case SEBuiltin(b) => (args.size <= b.arity)
+            case _: SEBuiltinRecursiveDefinition => (args.size <= 3)
```

Comment:
This constant `3` here looks like a ticking timebomb. It's not very unlikely that we need to add more recursive builtins in the future and some of them might have a different arity.

Reply:
Fixed